src/coreclr/jit/llvm.cpp

```diff
@@ -841,26 +852,131 @@ llvm::Value* Llvm::buildUserFuncCall(GenTreeCall* call)
     return mapGenTreeToValue(call, llvmCall);
 }
 
-void Llvm::buildCall(GenTree* node)
+llvm::FunctionType* Llvm::buildHelperLlvmFunctionType(GenTreeCall* call, bool withShadowStack)
 {
-    GenTreeCall* call = node->AsCall();
-    if (call->gtCallType == CT_HELPER)
+    Type* retLlvmType = getLlvmTypeForVarType(call->TypeGet());
+    std::vector<llvm::Type*> argVec;
+
+    if (withShadowStack)
+    {
+        argVec.push_back(Type::getInt8PtrTy(_llvmContext));
+    }
+
+    for (GenTreeCall::Use& use : call->Args())
     {
-        if (call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_STATIC_BASE))
+        Type* argLlvmType = getLlvmTypeForVarType(use.GetNode()->TypeGet());
+        argVec.push_back(argLlvmType);
+    }
+
+    return FunctionType::get(retLlvmType, ArrayRef<llvm::Type*>(argVec), false);
+}
+
+bool Llvm::helperRequiresShadowStack(CORINFO_METHOD_HANDLE corinfoMethodHnd)
+{
+    return corinfoMethodHnd == _compiler->eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE);
+}
+
+void Llvm::buildHelperFuncCall(GenTreeCall* call)
+{
+    if (call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_GENERIC_HANDLE) ||
+        call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_GENERIC_STATIC_BASE) ||
+        call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_DELEGATE_CTOR))
+    {
+        // TODO-LLVM
+        failFunctionCompilation();
+    }
+
+    if (call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_STATIC_BASE))
```

Comment:
For my education: what makes this case special?

Reply:
Probably the best thing I can do is point you to https://github.com/dotnet/runtimelab/issues/839 and https://github.com/dotnet/runtimelab/issues/1640.  

What this means in practice is that for most helpers,  `gtEntryPoint.handle` is a `nullptr` and for `CORINFO_HELP_READYTORUN_STATIC_BASE` there is no implementation in https://github.com/dotnet/runtimelab/blob/d623b2be57be73364ecf853a0fa31e7e9582b4ee/src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs#L240